Handlers (#790)
dotnet/docs-maui: docs/user-interface/handlers/customize.md

---
title: ".NET MAUI control customization with handlers"
description: "Learn how to customize .NET MAUI handlers, to augment the appearance and behavior of a cross-platform control."
ms.date: 08/15/2022
---

# Customize controls with handlers

[![Browse sample.](~/media/code-sample.png) Browse the sample](/samples/dotnet/maui-samples/userinterface-customizehandler)

> Handlers can be customized to augment the appearance and behavior of a cross-platform control beyond the customization that's possible through the control's API. This customization is achieved by modifying the mapper for a handler, with one of the following methods:

Reviewer: They augment the appearance and behavior of the cross-platform control or the native view control?

Author: I finally see your point on this. Ultimately it's augmenting the cross-platform control by modifying the native view.

- An `Action` that represents the method that performs the handler customization. The `Action` specifies two arguments:
  - A `handler` argument that provides an instance of the handler being customized.
  - A `view` argument that provides an instance of the cross-platform control that the handler implements.

> [!IMPORTANT]
> Handler customizations are global and aren't scoped to a specific control instance. Handler customization is allowed to happen anywhere in your app. Once a handler is customized, it affects all controls of that type, everywhere in your app.

> Each handler class exposes the native view that implements the cross-platform control via its `PlatformView` property. This property can be accessed to set native view properties, invoke native view methods, and subscribe to native view events. In addition, the cross-platform control implemented by the handler is exposed via its `VirtualView` property.

Reviewer: The native view implements the cross-platform control?? I'm confused 😢 I would think that it represents the cross-platform control, but it can't really implement it can it? I mean it's baked into the OS of the target platform right?

Handlers can be customized per platform by using conditional compilation, to multi-target code based on the platform. Alternatively, you can use partial classes to organize your code into platform-specific folders and files. For more information about conditional compilation, see [Conditional compilation](/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation).

## Customize a control

> The .NET MAUI `Entry` is a single-line text input control, that implements the `IEntry` interface. On iOS, the `EntryHandler` maps the `Entry` to an iOS `UITextField` native view. On Android, the `Entry` is mapped to an `AppCompatEditText` native view, and on Windows the `Entry` is mapped to a `TextBox` native view:

Reviewer: Would this be easier to read if it was a table or list?
```suggestion
The .NET MAUI `Entry` is a single-line text input control, that implements the `IEntry` interface. There is an `EntryHandler` for each platform, which maps the `Entry` control to the native view:

- iOS: `UITextFieldOn`
- Android: `AppCompatEditText`
- Windows: `TextBox`.
```

```xaml
<Entry HandlerChanged="OnEntryHandlerChanged"
       HandlerChanging="OnEntryHandlerChanging" />
```

Handlers can be customized per platform by using conditional compilation, or by using partial classes to organize your code into platform-specific folders and files. Each approach will be discussed in turn, by customizing an `Entry` so that all of its text is selected when it gains focus.

### Conditional compilation

The code-behind file containing the event handlers for the `HandlerChanged` and `HandlerChanging` events is shown in the following example, which uses conditional compilation:

```csharp
#if ANDROID
using AndroidX.AppCompat.Widget;
#elif IOS || MACCATALYST
using UIKit;
#elif WINDOWS
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml;
#endif

namespace CustomizeHandlersDemo.Views;

public partial class CustomizeEntryHandlerLifecyclePage : ContentPage
{
    public CustomizeEntryHandlerLifecyclePage()
    {
        InitializeComponent();
    }

    void OnEntryHandlerChanged(object sender, EventArgs e)
    {
#if ANDROID

>         ((sender as Entry).Handler.PlatformView as AppCompatEditText).SetSelectAllOnFocus(true);

Reviewer: This code causes the code box on the published page to scroll right. You could avoid that by casting at the start of the method
```csharp
Entry entry = sender as Entry;
```

The `OnEditingDidBegin` and `OnGotFocus` event handlers access the native view for the `Entry` on their respective platforms, and select all text that's in the `Entry`.

The `HandlerChanging` event is raised before the existing handler is removed from the cross-platform control, and before the new handler for the cross-platform control is created. Therefore, its event handler is where native event subscriptions should be removed, and other cleanup should be performed. The `HandlerChangingEventArgs` object that accompanies this event has `OldHandler` and `NewHandler` properties, which will be set to the old and new handlers respectively. In this example, the `OnHandlerChanging` event removes the subscription to the native view events on iOS, Mac Catalyst, and Windows.

### Partial classes

> Rather than using conditional compilation, it's also possible use partial classes to organize your control customization code into platform-specific folders and files. With this approach, your customization code is separated into a cross-platform partial class and a platform-specific partial class. The following example shows the cross-platform partial class:

Reviewer: minor fix
```suggestion
Rather than using conditional compilation, it's also possible to use partial classes to organize your control customization code into platform-specific folders and files. With this approach, your customization code is separated into a cross-platform partial class and a platform-specific partial class. The following example shows the cross-platform partial class:
```

```csharp
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace CustomizeHandlersDemo.Views
{
    public partial class CustomizeEntryPartialMethodsPage : ContentPage
    {
        partial void ChangedHandler(object sender, EventArgs e)
        {
            ((sender as Entry).Handler.PlatformView as TextBox).GotFocus += OnGotFocus;
        }

        partial void ChangingHandler(object sender, HandlerChangingEventArgs e)
        {
            if (e.OldHandler != null)
            {
                (e.OldHandler.PlatformView as TextBox).GotFocus -= OnGotFocus;
            }
        }

        void OnGotFocus(object sender, RoutedEventArgs e)
        {
            var nativeView = sender as TextBox;
            nativeView.SelectAll();
        }
    }
}
```

> The advantage of this approach is that conditional compilation isn't required, and that the partial methods don't have to be implemented on each platform. If an implementation isn't provided on a platform, then the method and all calls to the method are removed at compile time. For information about partial methods, see [Partial methods](/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods#partial-methods).

Reviewer: I feel like there should also be a link to the article that talks about platform specific code, folder layout, naming rules, and project settings.
